server/src/utils.ts

```diff
@@ -40,11 +39,41 @@ export let findProjectRootOfFile = (
       // reached top
       return null;
     } else {
-      return findProjectRootOfFile(dir);
+      return findProjectRootOfFileInDir(dir);
     }
   }
 };
 
+// TODO: races here?
+// TODO: this doesn't handle file:/// scheme
+export let findProjectRootOfFile = (
+  source: p.DocumentUri
+): null | p.DocumentUri => {
+  // First look in project files
+  let foundRootFromProjectFiles: string | null = null;
+
+  for (const rootPath of projectsFiles.keys()) {
```

Comment:
If I understand correctly here, the most common operation is to find the project info associated with the current open file.

I see you make a `{ [project root path]: [project info] }` index above, but it seems like it's missing the `{ [open file]: [project root path] }` or `{ [open file]: [project info] }` inverted index.

I think we don't need this loop if we're using an index for it in the same way

```ts
export const projectRoots = new Map<p.DocumentUri, string>();
export const projects = new Map<string, projectFiles>();
```

Reply:
We could do that, but this feels like a good trade off to me. That loop is cheap and it'll also be resilient to if roots change for some reason while the file is open (unlikely, but still).